lightning/src/ln/onion_message.rs

```diff
@@ -0,0 +1,69 @@
+/// The payload of an inbound onion message.
+pub struct Payload {
+	/// Onion messages may contain custom TLVs for the recipient.
+	custom_tlvs: Vec<CustomTlv>,
+	// Coming soon: path_id, reply_path
+}
+
+/// A sender, receiver and forwarder of onion messages. In upcoming releases, this object will be
+/// used to retrieve invoices and fulfill invoice requests from offers.
+pub struct OnionMessager<Signer: Sign, K: Deref>
+where K::Target: KeysInterface<Signer = Signer>
+{
+	keys_manager: K,
+	/// When `OnionMessager` receives onion messages for which our node is the final destination, it
+	/// delegates these messages' payloads to this handler. See [`Receive`] for more information.
+	inbound_handler: Receive,
+	pending_msg_events: Vec<MessageSendEvent>,
+}
+
+/// `Receive` defines behavior for receiving the final payload of an inbound onion message. These
+/// [`Payload`]s may contain a list of custom TLVs that are unknown to LDK, and it is up to the user
+/// how to handle them.
+///
+/// Until LDK implements BOLT 12 offers in upcoming releases, it is recommended to use
+/// [`IgnoringReceiver`] unless you have a specific need to receive custom onion messages.
+pub trait Receive {
+	/// Receive an inbound onion message payload.
+	fn receive(payload: Payload);
+}
+
+/// A [`Receive`]r that simply ignores inbound onion messages.
+pub struct IgnoringReceiver {}
+impl Receive for IgnoringReceiver {
+	fn receive(payload: Payload) {}
+}
+
+impl<Signer: Sign, K: Deref> OnionMessager<Signer, K>
+	where K::Target: KeysInterface<Signer = Signer>,
+{
+	/// Constructs a new `OnionMessager` to send, forward, and receive onion messages.
+	pub fn new(keys_manager: K) -> Self {}
+
+	/// Send an empty onion message to `recipient`, routing it through `intermediate_nodes`.
+	///
+	/// LDK will not make peer connections for you, so you must be connected to the first hop of
```

Comment:
That sounds weird at first but we're doing the same with channel establishment. So we expect a user to fetch the first `PublicKey` of intermediate node, query `NetworkGraph` to get a `NetAddress` and establish a connection ? Though I expect if the user failed to do so we're going to fail fast and return a `APIError::ConnectionUnavailable`.